src/vm/analysis/type_checker/natives/iterables.rs

```diff
@@ -0,0 +1,200 @@
+            _ => Err(CheckErrors::ExpectedListOrBuffer(argument_type.clone()))
+        }?;
+    
+        let filter_type = function_type.check_args(&[input_type.clone()])?;
+
+        if TypeSignature::BoolType != filter_type {
+            return Err(CheckErrors::TypeError(TypeSignature::BoolType, filter_type).into())
+        }
+    }
+
+    Ok(argument_type)
+}
+
+pub fn check_special_fold(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    check_argument_count(3, args)?;
+    
+    let function_name = args[0].match_atom()
+        .ok_or(CheckErrors::NonFunctionApplication)?;
+    // we will only lookup native or defined functions here.
+    //   you _cannot_ fold a special function.
+    let function_type = get_simple_native_or_user_define(function_name, checker)?;
+    
+    let argument_type = checker.type_check(&args[1], context)?;
+
+    let input_type = match argument_type {
+        TypeSignature::ListType(list_data) => Ok(list_data.destruct().0),
+        TypeSignature::BufferType(_) => Ok(TypeSignature::min_buffer()),
+        _ => Err(CheckErrors::ExpectedListOrBuffer(argument_type))
+    }?;
+
+    let initial_value_type = checker.type_check(&args[2], context)?;
+
+    // fold: f(A, B) -> A
+    //     where A = initial_value_type
+    //           B = list items type
+    
+    // f must accept the initial value and the list items type
+    let return_type = function_type.check_args(&[input_type.clone(), initial_value_type])?;
+
+    // f must _also_ accepts its own return type!
+    let return_type = function_type.check_args(&[input_type, return_type])?;
+    
+    Ok(return_type)
+}
+
+pub fn check_special_concat(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    check_argument_count(2, args)?;
+    
+    let lhs_type = checker.type_check(&args[0], context)?;
+    match lhs_type {
+        TypeSignature::ListType(lhs_list) => {
+            let rhs_type = checker.type_check(&args[1], context)?;
+            if let TypeSignature::ListType(rhs_list) = rhs_type {
+                if lhs_list.entry_type.admits_type(&*rhs_list.entry_type) {
+                    let return_type = TypeSignature::list_of(*lhs_list.entry_type, lhs_list.max_len + rhs_list.max_len)?;
+                    return Ok(return_type);
+                } else {
+                    return Err(CheckErrors::TypeError(*lhs_list.entry_type, *rhs_list.entry_type).into());
+                }
+            } else {
+                return Err(CheckErrors::TypeError(rhs_type.clone(), TypeSignature::ListType(lhs_list)).into());
+            }
+        },
+        TypeSignature::BufferType(lhs_buff_len) => {
+            let rhs_type = checker.type_check(&args[1], context)?;
+            if let TypeSignature::BufferType(rhs_buff_len) = rhs_type {
+                let size: u32 = u32::from(lhs_buff_len) + u32::from(rhs_buff_len);
```

Comment:
Here as well

Reply:
Or add a comment that makes clear that the overflow can't occur -- I'm pretty sure max buffer length * 2 cannot overflow u32

Reply:
Addressed with https://github.com/blockstack/blockstack-core/pull/1139/commits/5eb2a3e13c04c5ac846095bc0cd59c973c94fa48